pkg/controlplane/apiserver/server.go

```diff
@@ -212,10 +214,17 @@ func (c completedConfig) New(name string, delegationTarget genericapiserver.Dele
 				return nil
 			})
 		if c.Extra.PeerProxy != nil {
-			s.GenericAPIServer.AddPostStartHookOrDie("unknown-version-proxy-filter", func(context genericapiserver.PostStartHookContext) error {
+			// Wait for handler to be ready.
+			s.GenericAPIServer.AddPostStartHookOrDie("mixed-version-proxy-handler", func(context genericapiserver.PostStartHookContext) error {
 				err := c.Extra.PeerProxy.WaitForCacheSync(context.Done())
 				return err
 			})
+
+			// Run peer-discovery sync loop
+			s.GenericAPIServer.AddPostStartHookOrDie("peer-discovery-controller", func(context genericapiserver.PostStartHookContext) error {
+				go c.Extra.PeerProxy.RunPeerDiscoveryCacheSync(context, 1)
```

Comment:
Having a post start hook and running a goroutine doesn't really wait for anything. You can call `Run()` in a goroutine but separate out the `WaitForCacheSync()` so we can block readiness until informer caches are synced. (Similar to how `mixed-version-proxy-handler` is done)

Reply:
Alright moved all the waiting logic inside WaitForCacheSync() and have 2 separate goroutines for starting the local-discovery-cache-sync and peer-discovery-cache-sync.